[flight recorder] record NCCL commHash

shengbao-zheng opened the pull request:
Summary: NCCL commHash can be used to map the entry in flight recorder to NCCL trace dumped.  This PR includes using NCCL commHash when defining work as well as exposing commHash to flight recorder output.

Differential Revision: D54349909




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @rohan-varma

wconstab reviewed (commented):
this does look like the most straightforward way to do it, but i wonder if its worth being more efficient with the 'record size' -- is uid_ 1:1 with a commID?  if so, can we just put a mapping of uid (process group ID) <-> commhash into a separate table in the FR?

or is there more than one commHash per processgroup (e.g. for p2p vs allgather)

shengbao-zheng commented:
@wconstab 
> is uid_ 1:1 with a commID? if so, can we just put a mapping of uid (process group ID) <-> commhash into a separate table in the FR?
> or is there more than one commHash per processgroup (e.g. for p2p vs allgather)
No, it is not a 1:1 mapping between uid_ and commHash. commHash is per ncclComm. Thus in a p2p group, we will have two ncclComms, which means we actually need a list of commHash per PG in FR.

wconstab commented:
Ok makes sense to me. @shuqiangzhang @zdevito can finish the review while I'm out

shengbao-zheng committed:
[flight recorder] record NCCL commHash (#120975)

Summary:

NCCL commHash can be used to map the entry in flight recorder to NCCL trace dumped.  This PR includes using NCCL commHash when defining work as well as exposing commHash to flight recorder output.

Differential Revision: D54349909

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -323,8 +323,8 @@ std::string dump_nccl_trace() {
-    std::string ncclUniqueIDStr = buildNcclUniqueIdStr(ncclComm->getNcclId());
-    ncclDumpMap[ncclUniqueIDStr] = ncclComm->ncclCommDump();
+    auto commHash = ncclComm->getCommHash();
+    ncclDumpMap[commHash] = ncclComm->ncclCommDump();
```
commHash is a uint, but the map key is a string, this is implicit type conversion? better make the type explicit. Also the code needs to work even if NCCL_COMM_GET_UNIQUE_HASH is not defined. In which case, comhash would be 0, and every comm dump will have the same commHash and the dump would be inaccurate?

shengbao-zheng replied:
In ncclCommDump function we use commHash, thus the definition of NCCL_COMM_DUMP should always have NCCL_COMM_GET_UNIQUE_HASH defined, and commHash=0 should not be expected here. Also due to this, I changed the map key to uint.

shengbao-zheng replied:
NCCL_COMM_GET_UNIQUE_HASH is already defined in NCCL-EXP and we have some functions in both NCCL and PyTorch using commHash even if we don't do ncclCommDump. Thus it would be better to keep both macro: 1) when we do ncclCommDump, commHash will always be valid, 2) when we don't have NCCL_COMM_DUMP defined, functions using commHash can still work.

shuqiangzhang commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -394,6 +400,8 @@ struct NCCLTraceBuffer {
     // group)
     size_t seq_id_;
     size_t op_id_;
+    uint64_t nccl_comm_hash_; // commHash used inside NCCL to correlate with
```
Note flight recorder needs to work when we not using NCCL_EXP, in which case uid_ is more generic than comm_hash?

shengbao-zheng replied:
Thanks for pointing this out. I think when we are not using NCCL_EXP, no commHash and nccl dump will be available, thus we can remove them. I will make them only available when NCCL_EXP and NCCL_COMM_DUMP are defined.

shengbao-zheng commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -472,6 +472,17 @@ class NCCLComm {
   }
 
+  uint64_t getCommHash() {
+#ifdef NCCL_COMM_GET_UNIQUE_HASH
```
Yes. In NCCL_EXP

kwen2501 commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -472,6 +472,17 @@ class NCCLComm {
+    // query from NCCL when first time called
+    if (commHash_ == 0) {
+      C10D_NCCL_CHECK(
+          ncclCommGetUniqueHash(ncclComm_, &commHash_), c10::nullopt);
```
Unfortunately it might be difficult to accept APIs that's not public. 
We'd like functionalities in PyTorch's OSS code be accessible for all PyTorch users.

There are two solutions:
(1) Advocate and upstream the API to public NCCL;
(2) Establish an abstract interface between PyTorch and a "plugin". In the plugin, you can make this non-public API call.

kwen2501 replied:
Btw, the value of a ncclComm_t is actually an address pointer. So it should be unique enough and safe to be used as a "hash".

minsii replied:
> Btw, the value of a ncclComm_t is actually an address pointer (which points to the comm object). So it should be unique enough and safe to be used as a "hash".

Hi @kwen2501 , the pointer is a local ptr, which gives different value across rank. In contrast, we want to expose commHash which is a global unique id used by all ranks in a given NCCL communicator

minsii replied:
> Unfortunately it might be difficult to accept APIs that's not public. We'd like functionalities in PyTorch's OSS code be accessible for all PyTorch users.
> 
> There are three solutions: (1) Advocate and upstream the API to NCCL; (2) Establish an abstract interface between PyTorch and a "plugin". In the plugin, you can make this non-public API call; (3) "Unpack" the functionality of the API here, for example, in this case: `hash = std::hash(ncclComm_.comm)`

I agree that either (1) or (2) are good in general but it will take time. We can take them asynchronously.  To unblock, would it be acceptable if we guard the use of ncclCommGetUniqueHash by using the above NCCL_COMM_GET_UNIQUE_HASH macro, and make it no-op when the macro is not defined? 
E.g.
```
uint64_t getCommHash() {
#ifdef NCCL_COMM_GET_UNIQUE_HASH
   ncclCommGetUniqueHash(comm, &commHash_);
#endif
}

kwen2501 replied:
Hi @minsii thanks for the clarification.
> commHash which is a global unique id used by all ranks in a given NCCL communicator

I think `ncclUniqueId` serves this purpose. Can you get a has from it?
Per nccl.h, its structure is:
```
typedef struct { char internal[NCCL_UNIQUE_ID_BYTES]; } ncclUniqueId
```
so you can get a hash from `ncclUniqueId` or `ncclUniqueId.internal`.

minsii replied:
> Hi @minsii thanks for the clarification.
> 
> > commHash which is a global unique id used by all ranks in a given NCCL communicator
> 
> I think `ncclUniqueId` serves this purpose. Can you get a has from it? Per nccl.h, its structure is:
> 
> ```
> typedef struct { char internal[NCCL_UNIQUE_ID_BYTES]; } ncclUniqueId
> ```
> 
> so you can get a hash from `ncclUniqueId` or `ncclUniqueId.internal`.



> Hi @minsii thanks for the clarification.
> 
> > commHash which is a global unique id used by all ranks in a given NCCL communicator
> 
> I think `ncclUniqueId` serves this purpose. Can you get a has from it? Per nccl.h, its structure is:
> 
> ```
> typedef struct { char internal[NCCL_UNIQUE_ID_BYTES]; } ncclUniqueId
> ```
> 
> so you can get a hash from `ncclUniqueId` or `ncclUniqueId.internal`.

It would require us to expose the "hash" function inside NCCL to get the same hashed value at NCCL PG level v.s. NCCL internal. The goal for us to expose commHash is to have same id in NCCL PG and NCCL, which would help us quickly find corresponding log/trace cross stack.

Besides, uniqueId will not be different in the commSplit mode between parent comm v.s. child comm. `commHash` today also doesn't fully guarantee the uniqueness in such use case, but we can fix it in NCCL-EXP to make it unique. In contrast, it would be harder to support the commSplit case if we change to expose the hash function. Hope the design choice makes sense to you.

kwen2501 replied:
I understand your use case better now.

I wonder if we can use the value in `ncclComm_t` to identify whether a PG's comm is the same as the one being operated in NCCL. And then use `ncclUniqueId` to identify whether a comm is the same one across ranks?
(The two together should solve the global + local identification need).

My very first comment holds whether there is #ifdef or not.
The #ifdef makes sure it builds, but does not achieve the concept of General Availability.

From code maintenance point of view, if different organizations have their own "NCCL" API definitions and all upstream to pytorch, then our code management would become intractable. This practice would be more acceptable under forks of pytorch, such as "mepa/pytorch", "geegle/pytorch", "macrosoft/pytorch" (all names made up), but less under "pytorch/pytorch."

kwen2501 replied:
For this specific case: 

If a quick solution is needed, we can do (3):
- copy NCCL's internal hashing function here;
- write a comment that says "this is how NCCL hashes its communicator IDs in logs; to ease debugging, we prefer the same hash, thus copying NCCL's hash function here."

For a long-term solution in case this functionality is surely needed, I recommend contacting NCCL to create an access API.
